File: apps/web/lib/api/reports.ts
import { apiFetch } from "./client";
import type { SupportedCurrency } from "../currency";
import type {
  CategoryBreakdownItem,
  DashboardSummary,
  RecentRecord,
  SubcategoryBreakdownItem,
} from "./dashboard";

export type ReportPeriodPreset = "current_month" | "previous_month" | "custom";

export type ReportPeriodInput =
  | { period: "current_month" | "previous_month"; start?: never; end?: never }
  | { period: "custom"; start: string; end: string };

export type ReportPeriod = {
  preset: ReportPeriodPreset | null;
  start: string;
  end: string;
};

export type TrendPoint = {
  bucket: string;
  granularity: "day" | "month";
  income_minor: number;
  expense_minor: number;
  remaining_minor: number;
  currency: SupportedCurrency;
};

export type MerchantTotal = {
  merchant_name: string;
  total_minor: number;
  count: number;
  currency: SupportedCurrency;
};

export type TeamActivityItem = {
  user_id: string;
  display_name: string | null;
  records_created: number;
};

export type SpendingSummary = {
  total_income_minor: number;
  total_expenses_minor: number;
  remaining_balance_minor: number;
  top_category: {
    category_id: string | null;
    category_name: string;
    total_minor: number;
    currency: SupportedCurrency;
  } | null;
  trend_direction: "up" | "down" | "flat";
  currency: SupportedCurrency;
};

export type AiSummaryLocale = "en" | "ar";

export type AiSummaryResponse = {
  locale: AiSummaryLocale;
  text: string;
};

export type ReportResponse = {
  workspace_id: string;
  period: ReportPeriod;
  summary: DashboardSummary;
  category_breakdown: CategoryBreakdownItem[];
  income_category_breakdown: CategoryBreakdownItem[];
  spending_trend: TrendPoint[];
  top_merchants: MerchantTotal[];
  recent_records: RecentRecord[];
  team_activity: TeamActivityItem[];
  pending_review_count: number;
  spending_summary: SpendingSummary;
};

export type SubcategoryDrilldownResponse = {
  main_category_id: string;
  main_category_name: string;
  subcategory_breakdown: SubcategoryBreakdownItem[];
};

export async function getReport(
  workspaceId: string,
  period: ReportPeriodInput = { period: "current_month" },
) {
  const params = new URLSearchParams({ period: period.period });

  if (period.period === "custom") {
    params.set("start", period.start);
    params.set("end", period.end);
  }

  return apiFetch<ReportResponse>(`/workspaces/${workspaceId}/reports?${params.toString()}`);
}

export async function getSubcategoryBreakdown(
  workspaceId: string,
  mainCategoryId: string,
  period: ReportPeriodInput = { period: "current_month" },
) {
  const params = new URLSearchParams({ period: period.period });

  if (period.period === "custom") {
    params.set("start", period.start);
    params.set("end", period.end);
  }

  return apiFetch<SubcategoryDrilldownResponse>(
    `/workspaces/${workspaceId}/reports/category-breakdown/${mainCategoryId}/subcategories?${params.toString()}`,
  );
}

export async function requestAiSummary(
  workspaceId: string,
  period: ReportPeriodInput,
  locale: AiSummaryLocale,
) {
  return apiFetch<AiSummaryResponse>(`/workspaces/${workspaceId}/reports/ai-summary`, {
    method: "POST",
    body: JSON.stringify({ ...period, locale }),
  });
}

File: apps/api/app/schemas/reports.py
from datetime import date
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.schemas.dashboard import CategoryBreakdownItem, FinancialSummary, RecentRecord


class ReportPreset(StrEnum):
    CURRENT_MONTH = "current_month"
    PREVIOUS_MONTH = "previous_month"
    CUSTOM = "custom"


class TrendGranularity(StrEnum):
    DAY = "day"
    MONTH = "month"


class TrendDirection(StrEnum):
    UP = "up"
    DOWN = "down"
    FLAT = "flat"


class ReportPeriod(BaseModel):
    preset: ReportPreset | None = None
    start: date
    end: date


class TrendPoint(BaseModel):
    bucket: date
    granularity: TrendGranularity
    income_minor: int
    expense_minor: int
    remaining_minor: int
    currency: str = "SAR"


class MerchantTotal(BaseModel):
    merchant_name: str
    total_minor: int
    count: int
    currency: str = "SAR"


class TeamActivityItem(BaseModel):
    user_id: UUID
    display_name: str | None = None
    records_created: int


class TopCategorySummary(BaseModel):
    category_id: UUID | None = None
    category_name: str
    total_minor: int
    currency: str = "SAR"


class SpendingSummary(BaseModel):
    total_income_minor: int
    total_expenses_minor: int
    remaining_balance_minor: int
    top_category: TopCategorySummary | None = None
    trend_direction: TrendDirection = TrendDirection.FLAT
    currency: str = "SAR"


class ReportData(BaseModel):
    workspace_id: UUID
    period: ReportPeriod
    summary: FinancialSummary
    category_breakdown: list[CategoryBreakdownItem]
    spending_trend: list[TrendPoint]
    top_merchants: list[MerchantTotal]
    recent_records: list[RecentRecord]
    team_activity: list[TeamActivityItem]
    pending_review_count: int
    spending_summary: SpendingSummary

    model_config = ConfigDict(from_attributes=True)
